### DISCUSSION TITLE
Access Validation during Replication

### DISCUSSION THREAD

**Discussion by spetrac** [DOME-Marketplace/architecture#7] [ADR]
## Context

We discussed different options how to handle the validation of entities before they will be replicated. This discussion came up because of the introduction of certain policies into the product offerings to control whether an offering is allowed to be replicated to another access node depending on its operator.
The problem we faced was that a policy on a product offering is its own entity during replication and even if the associated product offering could easily be validated, it is difficult to validate all the entities that might be connected somehow to that offering. The discussion led to validating of the entities by searching all its references, but the effort could be exponential with arbitrary references between entities.  
We figured that we would have to use the entity type to exactly specify how an entity should be evaluated. For this, we came up with a multiple step plan to iteratively adopt the validation of the different entities.

## Step 0

The validation of the policies is a task on its own, so we will start with creating the engine to interpret the operator restrictions. This will be a sub-type of the already existing product offering term. During replication we will then look **only** for the product offerings and restrict the replication to this entity type, even if other entities are connected.

The following documents should be used as reference for the policy engine:

- [OperatorRestriction.schema.json](https://github.com/DOME-Marketplace/dome-odrl-profile/blob/main/schemas/simplified/OperatorRestriction.schema.json)
- [Policy Evaluation Strategy - Operator Restriction](https://github.com/DOME-Marketplace/dome-odrl-profile/blob/main/docs/policy-evaluation-strategy.md#operator-restriction)
- [Catalogue visibility control - Restrict offerings for specific marketplaces](https://github.com/DOME-Marketplace/dome-odrl-profile/blob/main/docs/catalogue-visibility-control.md#catalogue-visibility-control)

## Step 1

Each entity type must be validated, but we recognize that some entities does not need protection, because they are shared between entities and does not contain any critical information. For this reason, these entities will have a pass-always rule configured, but all unknown entities will now be restricted from replication regardless.  
To reduce the effort from the previous step, all currently used entities will get the pass-always rule, except the product offering and the product offering term. The last two will be validated by the policy engine from step 0.

We have identified the following entity types to be currently used:

- [Individual](https://tmforum-rand.github.io/schemas/EngagedParty/Individual.schema.json)
- [Organization](https://tmforum-rand.github.io/schemas/EngagedParty/Organization.schema.json)
- [Catalog](https://tmforum-rand.github.io/schemas/Product/Catalog.schema.json)
- [Category](https://tmforum-rand.github.io/schemas/Product/Category.schema.json)
- [Product](https://tmforum-rand.github.io/schemas/Product/Product.schema.json)
- [ProductSpecification](https://tmforum-rand.github.io/schemas/Product/ProductSpecification.schema.json)
- [ProductOffering](https://tmforum-rand.github.io/schemas/Product/ProductOffering.schema.json)
- [ProductOfferingPrice](https://tmforum-rand.github.io/schemas/Product/ProductOfferingPrice.schema.json)
- [ResourceSpecification](https://tmforum-rand.github.io/schemas/Resource/ResourceSpecification.schema.json)
- [ServiceSpecification](https://tmforum-rand.github.io/schemas/Service/ServiceSpecification.schema.json)
- _([ProductOfferingTerm](https://tmforum-rand.github.io/schemas/Product/ProductOfferingTerm.schema.json))_

## Step 2

We will further identify the entity types that are meant to be protected alongside the product offering, e.g. the product specification or the product offering price. For this we will ask partners and work package owners to help with the decision and we will supply them with information about relevant attributes of the different entity types.  
When we have identified the relevant entity types, we will figure out strategies for each entity type individually to identify the relevant policies for its replication evaluation. E.g. the product specification could look at all referring product offerings and if one of them allows the replication, the product specification would also be replicated. This approach solves a problem were one entity might be linked by multiple other protected entities with different replication rules, but must not be dismissed if a valid entity relies on it.  
Using different evaluation strategies for the different entity types is feasible from our perspective, because the number of different entity types is quite small.

_(This step can also lead to possible future evaluation rules if there will be new use-cases for replicated entities like orders or bills.)_

**Comment by spetrac** [comment]
Regarding the policy schemas, the semantics of the marketplace restriction should maybe be changed from referring to marketplace to referring to access node operators, because not only marketplaces operate access nodes. Those access nodes will be in the list of trusted access nodes for a replication to be permitted.

- rename `permittedMarketplace` to `permittedOperator`
- rename `prohibitedMarketplace` to `prohibitedOperator`

Also fix descriptions of the schemas.

**Comment by spetrac** [comment]
Implementation approaches for policies:

- [iSHARE Authorization Registry](https://dev.ishare.eu/authorisation-registry-role/getting-started)
- [Trusted Policies Registry API](https://hub.ebsi.eu/apis/pilot/trusted-policies-registry)
- [OPA based ODRL PAP](https://github.com/wistefan/odrl-pap)
- [TMForum Policy Management API](https://www.tmforum.org/oda/open-apis/directory/policy-management-api-TMF723)

**Reply by wistefan** [reply]
The solution developed as part of DOME should also be considered: https://github.com/wistefan/odrl-pap

**Reply by spetrac** [reply]
The TMForum Policy Management API is also a candidate, but I cannot find any resources describing it:
https://www.tmforum.org/oda/open-apis/directory/policy-management-api-TMF723

**Comment by spetrac** [comment]
As Jesus reminded us, there are two requirements:

- **REQUIREMENT 005:** Every Product Offering is cryptographically associated to the Party (organisation) which published the Product Offering initially.
- **REQUIREMENT 006:** Every Product Offering which is Launched is cryptographically associated to the unique identity of the Access Node where is was launched initially.

And I remember the discussion to have the product offering as verifiable credential. For me, this translate to: The ProductOffering will be bundled with all associated entities into a VC and signed by the access node. This also means, the ProductOffering has to contain all associated policies or at least those that belong to the offering itself. All of this should somehow be compatible with the existing API and the existing entities. This is why I heavily doubt that we should use any solution that is not easily able to integrate with the existing framework and not semantically compatible with the existing entities.
